python/python/lance/dataset.py

```diff
@@ -1516,7 +1516,12 @@ def create_index(
             kwargs["shuffle_partition_concurrency"] = shuffle_partition_concurrency
 
         self._ds.create_index(column, index_type, name, replace, kwargs)
-        return LanceDataset(self.uri, index_cache_size=index_cache_size)
+
+        storage_options = None
+        if "storage_options" in kwargs:
+            storage_options = kwargs.pop("storage_options")
+
+        return LanceDataset(self.uri, index_cache_size=index_cache_size, storage_options=storage_options)
```

Comment:
`self._ds.create_index()` just updates self here:

https://github.com/lancedb/lance/blob/874673733698184c4b9772075eb3111c0dda7b6e/python/src/dataset.rs#L928-L934

So, we might as well do `return self` instead. The current code re-opens the dataset, but there's no good reason we can't re-use the instance given we are already mutating it. This will make the `index_cache_size` parameter un-used, but I think we can just deprecate it in that case.

Reply:
agree - I find it pretty weird there is a recreation in the first place

from an end-user perspective the current implementation is a blocker for usage of indices in meaningful fashions (not just local fs) so the "dirty" fix was implemented. since the `set_index` method is featured in the readme, should at least function with all object storages though

Reply:
> the current implementation is a blocker for usage of indices

Why is this? Why can't you just use the original dataset?

```python
ds = lance.write_dataset(...)
ds.create_index(...) # ignore return value
# use ds now?
```